floor and ceiling for DTR

emileten opened the pull request:
- [x] helps closing (but does not close) : 

- #468, #470, #472, #473, #474, #475, #476, #477, #481 from https://github.com/ClimateImpactLab/downscaleCMIP6/issues

- [x] tests added / passed
- [x] docs reflect changes
- [x] entry in CHANGELOG.md

- rename the DTR correction core function to 'floor...'
- added an equivalent one for a ceiling only applied to [-60, 60]. Default ceiling is 70. 
-  merged these two functionalities in one service renamed to correct_dtr
- added tests for the latter 
-  accordingly modified click 
- changed `<` to `<=` in the DTR range validation 

This means our DTR 'validation' looses its meaning from when the DTR correction is applied on. It's a quick-and-dirty fix. 

@dgergel this is just applying a cap at 70. This follows what I saw in the data (data points away from 60th north parallel but with a value significantly larger than 70) and me realizing (1) that 60th north parallel already crosses a significant european population area and (2) 70 diff in kelvin is already quite extreme

brews commented:
So we can keep track of it, can someone list out the downscaleCMIP6 issues this is related to? This will create links and it will be easier to keep track of what needs to be rerun.

I see https://github.com/ClimateImpactLab/downscaleCMIP6/issues/476. What else?

brews commented on dodola/services.py:
```diff
@@ -641,21 +642,24 @@ def correct_wet_day_frequency(x, out, process):
-def correct_small_dtr(x, out, threshold=1.0):
-    """Corrects small diurnal temperature range (DTR) values in a dataset
+def correct_dtr(x, out, floor=1.0, ceiling=70.0):
+    """Corrects diurnal temperature range (DTR) values in a dataset
```
We need to describe what's actually happening aside from a "correction". It's a bit more complex now. At least outline the main components.

This is a good place for a one-line summary at the top of the docstr and then an extended description below it:

```python
""""This is a one line-summary of what the function does.

This is an extended description talking about things beyond one line and going into
gross detail about all the methods and details and all that stuff. If we needed more 
than one line for the summary description, its usually a sign that our function was trying 
to do too much.

Parameters
-----------
...
"""
```

I can live without extensive documentation on the CLI but we need to cover ouselves better in `dodola.services` and `dodola.core`. I feel like we're going to come back to this in a month with a poor idea about what all this actually does.

emileten replied:
Yes that totally makes sense. I wrote this too fast 👍

brews commented on CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 - Update wet day frequency correction to incorporate method additions from Hempel et al 2013. (PRs #162 and #159, @dgergel)
+- floor and ceiling for DTR (PR #163 @emileten)
```
Would you mind putting these two items under a `## Changed` heading here?

brews replied:
Nevermind! I just added this heading to `main`. You will likely need to resolve the merge conflict though (sorry).

Marking resolved.

brews reviewed (approved):
@emileten I think this looks good. I have two quick comments on expanding documentation. Aside from these I think we can merge this and take things for a run. Thanks!

brews commented:
I took a look at this with a programmer's eye. @dgergel, I'd appreciate it if you could review this and double-check the method actually does what you want/described.

dgergel commented on dodola/services.py:
```diff
@@ -641,21 +642,24 @@ def correct_wet_day_frequency(x, out, process):
     ds = storage.read(x)
-    ds_corrected = apply_small_dtr_correction(ds, threshold)
-    storage.write(out, ds_corrected)
+    ds = apply_dtr_floor(ds, floor)
```
I'd make the application of these conditional on the `correction_type` argument

dgergel reviewed (changes requested):
@emileten this looks really good. Nice work! 

My one suggestion for change - which should be really quick - is adding a `correction_type` argument to the CLI with options `ceiling`, `floor` or `both` (or something like that). Even though we're using dodola core functions in our workflow, since we're applying the ceiling and floor functionality at different points in the workflow, I think it'd be nice to have the option to only apply one here and have that be a service.

emileten commented on dodola/cli.py:
```diff
@@ -573,9 +573,11 @@ def correct_wetday_frequency(x, out, process):
     "--threshold", "-t", help="Threshold for correcting small DTR values up to"
 )
-def correct_dtr(x, out, threshold=1.0):
+def correct_dtr(x, out, floor=1.0, ceiling=70.0):
```
Thanks @dgergel just realized that. 

I might implement that with simply allowing the `None` value for either and set that to default to minimize the amount of new code.

emileten commented:
@dgergel @brews I addressed all your concerns and following our conversation, structurally modified things a bit. 

`floor` and `ceiling` are two separate functions, in `services` and `cli` as well. That keeps things clean and there is no step when we're doing both. 

@brews I am modifying the header of this to include all related issues.

dgergel reviewed (approved):
LGTM!

brews reviewed (approved):
@emileten Technically, I think this revision is fine. The changelog and PR description are a bit terse, but that's a minor comment given the deadline you're under for this.

brews commented:
@emileten BTW, I think you can merge this whenever you're ready!
